build-support/bin/generate_completions.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import json
+import subprocess
+
+BASH_COMPLETION_TEMPLATE = """
+# DO NOT EDIT.
+# This script is autogenerated by build-support/bin/generate_autocomplete.py
+
+readonly PANTS_GOALS="\\
+    {goals}"
+
+{goal_options}
+
+function pants_completions()
+{{
+    local current_word previous_word previous_goal
+
+    current_word=${{COMP_WORDS[COMP_CWORD]}}
+    previous_word=${{COMP_WORDS[COMP_CWORD-1]}}
+    previous_goal=$(get_previous_goal)
+
+    case "$previous_goal" in
+    {goal_cases}
+    esac
+
+    # Show goals, unless the tab completion has a hyphen - then show global/goal options instead
+    if [[ $current_word == -* ]]; then
+        COMPREPLY=( $( compgen -W "$complete_options" -- $current_word ))
+    else
+        COMPREPLY=( $( compgen -W "$PANTS_GOALS" -- $current_word ))
+    fi
+    return 0
+}}
+
+# Get the most recent goal in the command arguments, so options can be correctly applied
+# This function will ignore hyphenated options when looking for the goal
+# If this is empty, we're at the top-level Pants command
+function get_previous_goal()
+{{
+    local previous_goal i current_word
+    previous_goal=
+    for (( i=$COMP_CWORD; i > 0; --i )); do
+        current_word=${{COMP_WORDS[i]}}
+        if [[ $current_word != -* ]]; then
+            previous_goal=$current_word
+            break
+        fi
+    done
+    echo $previous_goal
+}}
+
+# TODO: Need to create a separate .zsh script instead of this
+autoload -U +X bashcompinit && bashcompinit
+autoload -Uz compinit && compinit
+
+complete -F pants_completions pants
+"""
+
+GOAL_OPTIONS_TEMPLATE = """
+readonly PANTS_{name}_OPTIONS="\\
+    {options}"
+"""
+
+GOAL_CASE_TEMPLATE = """
+    {goal})
+        complete_options="$PANTS_{bash_name}_OPTIONS"
+        ;;
+"""
+
+
+def _run_pants_help():
+    process = subprocess.run(["./pants", "help-all"], capture_output=True, check=True)
```

Comment:
What should I be calling here if I want to run this from `./pants run`? I can't pants-in-pants from what I've seen.

Previously, I used `run_pants` from testutils, but it seems like I need to put in my own packages, which knocks it out of alignment with the current workspace. 

I spent a long time trying to get the Scheduler working, but eventually abandoned that in order to get the rest of this ticket completed.

Reply:
@Eric-Arellano Bump? Not sure what the best pattern is here for pants-in-pants calls

Will fall back to just a regular python script if needed, not a huge problem today.

Reply:
Yeah, pants-in-pants isn't really a thing. The idiom would be to make this a pants goal and have it operate directly on the `AllHelpInfo` object, instead of the roundabout way of generating the JSON and then parsing it.

Reply:
Will be in a new PR once we decide on a goal